Source/Core/Core/IOS/FS/HostBackend/FS.cpp

```diff
@@ -252,99 +255,135 @@ HostFileSystem::FstEntry* HostFileSystem::GetFstEntryForPath(const std::string&
+  File::FSTEntry parent_entry = File::ScanDirectoryTree(Path, true);
+  std::deque<File::FSTEntry> todo;
+  todo.insert(todo.end(), parent_entry.children.begin(), parent_entry.children.end());
 
-    while (!todo.empty())
+  while (!todo.empty())
+  {
+    File::FSTEntry& entry = todo.front();
+    std::string name = entry.physicalName;
+    name.erase(0, Path.length() + 1);
+    char type = entry.isDirectory ? 'd' : 'f';
+    p.Do(type);
+    p.Do(name);
+    if (entry.isDirectory)
+    {
+      todo.insert(todo.end(), entry.children.begin(), entry.children.end());
+    }
+    else
     {
-      File::FSTEntry& entry = todo.front();
-      std::string name = entry.physicalName;
-      name.erase(0, Path.length() + 1);
-      char type = entry.isDirectory ? 'd' : 'f';
-      p.Do(type);
-      p.Do(name);
-      if (entry.isDirectory)
+      u32 size = (u32)entry.size;
+      p.Do(size);
+
+      File::IOFile handle(entry.physicalName, "rb");
+      char buf[BUFFER_CHUNK_SIZE];
+      u32 count = size;
+      while (count > BUFFER_CHUNK_SIZE)
       {
-        todo.insert(todo.end(), entry.children.begin(), entry.children.end());
+        handle.ReadArray(&buf[0], BUFFER_CHUNK_SIZE);
+        p.DoArray(buf);
+        count -= BUFFER_CHUNK_SIZE;
       }
-      else
+      handle.ReadArray(&buf[0], count);
+      p.DoArray(&buf[0], count);
+    }
+    todo.pop_front();
+  }
+
+  char type = 0;
+  p.Do(type);
+}
+
+void HostFileSystem::DoState(PointerWrap& p)
+{
+  // Temporarily close the file, to prevent any issues with the savestating of files/folders.
+  for (Handle& handle : m_handles)
+    handle.host_file.reset();
+
+  bool originalSaveStateMadeDuringMovieRecording = Movie::IsMovieActive();
+  p.Do(originalSaveStateMadeDuringMovieRecording);
+
+  u32 tempVal = 0;
+
+  if (!p.IsReadMode())
+  {
+    DoStateWriteOrMeasure(p, "/tmp");
+    u8* prevPosition = p.GetPositionAndDoInt(tempVal);
+    if (Movie::IsMovieActive())
+    {
+      DoStateWriteOrMeasure(p, "/");
+      if (p.IsWriteMode())
       {
-        u32 size = (u32)entry.size;
-        p.Do(size);
-
-        File::IOFile handle(entry.physicalName, "rb");
-        char buf[65536];
-        u32 count = size;
-        while (count > 65536)
-        {
-          handle.ReadArray(&buf[0], 65536);
-          p.DoArray(buf);
-          count -= 65536;
-        }
-        handle.ReadArray(&buf[0], count);
-        p.DoArray(&buf[0], count);
+        u32 sizeOfNand = p.GetOffsetFromPreviousPosition(prevPosition);
+        memcpy(prevPosition, &sizeOfNand, sizeof(u32));
       }
-      todo.pop_front();
     }
+  }
 
-    char type = 0;
-    p.Do(type);
+  else  // case where we're in read mode.
+  {
+    DoStateRead(p, "/tmp");
+    if (originalSaveStateMadeDuringMovieRecording != Movie::IsMovieActive())
+    {
+      u8* temp = p.DoExternal(tempVal);
+      if (temp == 0)  // this branch just exists to prevent a compiler error.
+        return;
```

Comment:
I don't understand what this is doing...?

Reply:
temp contains the result of DoExternal(). Since the DoExternal() function is annotated with the [[nodiscard]] annotation, Visual Studio will throw a compiler error if the value returned by the function is never read. This useless if statement prevents that error.

Reply:
Ah, because of the nodiscard. The convention is that you explicitly acknowledge discarding a nodiscard result by casting the result to void, ie. `(void)p.DoExternal(val);`. That'll make the compiler not count it as a warning either.